batcher/aligned-sdk/src/communication/messaging.rs

```diff
@@ -0,0 +1,213 @@
+use futures_util::{future, stream::SplitStream, SinkExt, StreamExt, TryStreamExt};
+use log::{debug, error};
+use std::{collections::HashSet, sync::Arc};
+use tokio::{net::TcpStream, sync::Mutex};
+
+use ethers::{core::k256::ecdsa::SigningKey, signers::Wallet};
+use futures_util::stream::SplitSink;
+use tokio_tungstenite::{tungstenite::Message, MaybeTlsStream, WebSocketStream};
+
+use crate::{
+    communication::batch::{
+        handle_batch_inclusion_data, handle_batch_inclusion_data_without_await,
+    },
+    core::{
+        errors::SubmitError,
+        types::{
+            AlignedVerificationData, ClientMessage, ResponseMessage, VerificationData,
+            VerificationDataCommitment,
+        },
+    },
+    eth::{aligned_service_manager, BatchVerifiedEventStream, BatchVerifiedFilter},
+};
+
+pub async fn send_messages(
+    ws_write: Arc<Mutex<SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>>>,
+    verification_data: &[VerificationData],
+    wallet: Wallet<SigningKey>,
+) -> Result<Vec<VerificationData>, SubmitError> {
+    let mut sent_verification_data = Vec::new();
+
+    let mut ws_write = ws_write.lock().await;
+
+    for verification_data in verification_data.iter() {
+        let msg = ClientMessage::new(verification_data.clone(), wallet.clone()).await;
+        let msg_str = serde_json::to_string(&msg).map_err(SubmitError::SerdeError)?;
+        ws_write
+            .send(Message::Text(msg_str.clone()))
+            .await
+            .map_err(SubmitError::ConnectionError)?;
+        sent_verification_data.push(verification_data.clone());
+        debug!("Message sent...");
+    }
+
+    Ok(sent_verification_data)
+}
+
+pub async fn receive(
+    ws_read: SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>,
+    ws_write: Arc<Mutex<SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>>>,
+    total_messages: usize,
+    num_responses: Arc<Mutex<usize>>,
+    verification_data_commitments_rev: &mut Vec<VerificationDataCommitment>,
+) -> Result<Option<Vec<AlignedVerificationData>>, SubmitError> {
+    // Responses are filtered to only admit binary or close messages.
+    let mut response_stream =
+        ws_read.try_filter(|msg| future::ready(msg.is_binary() || msg.is_close()));
+
+    let mut aligned_verification_data: Vec<AlignedVerificationData> = Vec::new();
+
+    while let Some(Ok(msg)) = response_stream.next().await {
+        if let Message::Close(close_frame) = msg {
+            if let Some(close_msg) = close_frame {
+                error!("Connection was closed before receiving all messages. Reason: {}. Try submitting your proof again", close_msg.to_owned());
+                ws_write.lock().await.close().await?;
+                return Ok(None);
+            }
+            error!("Connection was closed before receiving all messages. Try submitting your proof again");
+            ws_write.lock().await.close().await?;
+            return Ok(None);
```

Comment:
same here

Reply:
Solved in [cef4afb](https://github.com/yetanotherco/aligned_layer/pull/583/commits/cef4afb1e728da34f45728b42d6780e29a862585)